
<launch>
    
    <!--- Simulate Time Parameters -->    
    <arg name = "sim"    value = "false" />
    <arg name = "rviz_p" value = "true" />
    <param name = "/use_sim_time" value = "false" /> <!-- use true for simulation simGetSimulationTime -->

    <!--- param -->
    <rosparam command="load" file="$(find ekf_loam)/config/ekf_loam_parameters.yaml" /> 
    <!--rosparam unless="$(arg sim)" command="load" file="$(find ekf_loam)/config/os1_16_parameters.yaml" /-->
    <rosparam unless="$(arg sim)" command="load" file="$(find ekf_loam)/config/velodyne_16_parameters.yaml" />
    <rosparam if="$(arg sim)" command="load" file="$(find ekf_loam)/config/vlp_16_sim_parameters.yaml" />
    <rosparam command="load" file="$(find adaptive_filter)/config/adaptive_filter_parameters.yaml" />
    <!--rosparam command="load" file="$(find adaptive_filter)/config/nfn_parameters.yaml" /-->

    <!-- TFs -->
    <include file="$(find ekf_loam)/launch/ekf_loam_tf.launch" />  

    <!-- odom 
    <node pkg="espeleo_locomotion" type="odom.py" name="odom" output="screen" /> 
    <rosparam file="$(find espeleo_locomotion)/config/locomotion_parameters.yaml" command="load" ns=""/>  -->

    <!--- EKF-LOAM --> 
    <node pkg="ekf_loam" 
          type="PreTreatment"    
          name="PreTreatment2"    
          output="screen">
	<!--<remap from="/os1_cloud_node/points" to="/velodyne/points2"/> 
        <remap from="/os1_cloud_node/points" to="/os1_cloud_node/points"/> -->
        <remap from="/os1_cloud_node/points" to="/velodyne_points"/>
  	<remap from="/ekf_loam/full_cloud_projected" to="/ekf_loam2/full_cloud_projected"/>
  	<remap from="/ekf_loam/full_cloud_info" to="/ekf_loam2/full_cloud_info"/>
  	<remap from="/ekf_loam/ground_cloud" to="/ekf_loam2/ground_cloud"/>
  	<remap from="/ekf_loam/segmented_cloud" to="/ekf_loam2/segmented_cloud"/>
  	<remap from="/ekf_loam/segmented_cloud_pure" to="/ekf_loam2/segmented_cloud_pure"/>
  	<remap from="/ekf_loam/segmented_cloud_info" to="/ekf_loam2/segmented_cloud_info"/>
  	<remap from="/ekf_loam/outlier_cloud" to="/ekf_loam2/outlier_cloud"/>
  	<remap from="/ekf_loam/PreTreatmentTime" to="/ekf_loam2/PreTreatmentTime"/>
    </node>
    <node pkg="ekf_loam" 
          type="LiDAROdometry" 
          name="LiDAROdometry2" 
          output="screen">
  	<remap from="/ekf_loam/segmented_cloud" to="/ekf_loam2/segmented_cloud"/>
  	<remap from="/ekf_loam/segmented_cloud_info" to="/ekf_loam2/segmented_cloud_info"/>
  	<remap from="/ekf_loam/outlier_cloud" to="/ekf_loam2/outlier_cloud"/>
  	<remap from="/ekf_loam/PreTreatmentTime" to="/ekf_loam2/PreTreatmentTime"/>
  	<remap from="/ekf_loam/laser_cloud_sharp" to="/ekf_loam2/laser_cloud_sharp"/>
  	<remap from="/ekf_loam/laser_cloud_less_sharp" to="/ekf_loam2/laser_cloud_less_sharp"/>
  	<remap from="/ekf_loam/laser_cloud_flat" to="/ekf_loam2/laser_cloud_flat"/>
  	<remap from="/ekf_loam/laser_cloud_less_flat" to="/ekf_loam2/laser_cloud_less_flat"/>
  	<remap from="/ekf_loam/laser_cloud_corner_last" to="/ekf_loam2/laser_cloud_corner_last"/>
  	<remap from="/ekf_loam/laser_cloud_surf_last" to="/ekf_loam2/laser_cloud_surf_last"/>
  	<remap from="/ekf_loam/outlier_cloud_last" to="/ekf_loam2/outlier_cloud_last"/>
  	<remap from="/ekf_loam/laser_odom_to_init" to="/ekf_loam2/laser_odom_to_init"/>
  	<remap from="/ekf_loam/LiDAROdometryTime" to="/ekf_loam2/LiDAROdometryTime"/>
  	<remap from="/ekf_loam/LiDAROdometryTimeA" to="/ekf_loam2/LiDAROdometryTimeA"/>
    </node> 
    <node pkg="adaptive_filter" 
          type="EKFAdaptiveFilter"     
          name="EKFAdaptiveFilter2"     
          output="screen">
  	<remap from="/ekf_loam/laser_odom_to_initOut" to="/ekf_loam2/laser_odom_to_initOut"/>
  	<remap from="/ekf_loam/filter_odom_to_init" to="/ekf_loam2/filter_odom_to_init"/>
    </node> 
    <node pkg="ekf_loam" 
	  type="TransformOdometry" 
	  name="TransformOdometry_to_filter2" 
	  args="foward" 
	  output="screen" >
  	<remap from="/ekf_loam/laser_odom_to_init" to="/ekf_loam2/laser_odom_to_init"/>
  	<remap from="/ekf_loam/laser_odom_to_initOut" to="/ekf_loam2/laser_odom_to_initOut"/>
  	<remap from="/ekf_loam/filter_odom_to_init" to="/ekf_loam2/filter_odom_to_init"/>
  	<remap from="/ekf_loam/filter_odom_to_initOut" to="/ekf_loam2/filter_odom_to_initOut"/>
    </node>         
    <node pkg="ekf_loam" 
	  type="TransformOdometry" 
	  name="TransformOdometry_to_loam2" 
	  args="back" 
	  output="screen" >
  	<remap from="/ekf_loam/laser_odom_to_init" to="/ekf_loam2/laser_odom_to_init"/>
  	<remap from="/ekf_loam/laser_odom_to_initOut" to="/ekf_loam2/laser_odom_to_initOut"/>
  	<remap from="/ekf_loam/filter_odom_to_init" to="/ekf_loam2/filter_odom_to_init"/>
  	<remap from="/ekf_loam/filter_odom_to_initOut" to="/ekf_loam2/filter_odom_to_initOut"/>
    </node>
    <node pkg="ekf_loam" 
          type="LiDARMapping"     
          name="LiDARMapping2"     
          output="screen">
  	<remap from="/ekf_loam/laser_cloud_corner_last" to="/ekf_loam2/laser_cloud_corner_last"/>
  	<remap from="/ekf_loam/laser_cloud_surf_last" to="/ekf_loam2/laser_cloud_surf_last"/>
  	<remap from="/ekf_loam/outlier_cloud_last" to="/ekf_loam2/outlier_cloud_last"/>
  	<remap from="/ekf_loam/LiDAROdometryTimeA" to="/ekf_loam2/LiDAROdometryTimeA"/>
  	<remap from="/ekf_loam/filter_odom_to_initOut" to="/ekf_loam2/filter_odom_to_initOut"/>
  	<remap from="/ekf_loam/laser_odom_to_init" to="/ekf_loam2/laser_odom_to_init"/>
  	<remap from="/ekf_loam/key_pose_origin" to="/ekf_loam2/key_pose_origin"/>
  	<remap from="/ekf_loam/laser_cloud_surround" to="/ekf_loam2/laser_cloud_surround"/>
  	<remap from="/ekf_loam/aft_mapped_to_init" to="/ekf_loam2/aft_mapped_to_init"/>
  	<remap from="/ekf_loam/history_cloud" to="/ekf_loam2/history_cloud"/>
  	<remap from="/ekf_loam/corrected_cloud" to="/ekf_loam2/corrected_cloud"/>
  	<remap from="/ekf_loam/recent_cloud" to="/ekf_loam2/recent_cloud"/>
  	<remap from="/ekf_loam/mapOptmizationTime" to="/ekf_loam2/mapOptmizationTime"/>
  	<remap from="/ekf_loam/mapOptmizationTimeA" to="/ekf_loam2/mapOptmizationTimeA"/>
  	<remap from="/ekf_loam/mapOptmizationTimeM" to="/ekf_loam2/mapOptmizationTimeM"/>
  	<remap from="/ekf_loam/mapOptmizationTimeL" to="/ekf_loam2/mapOptmizationTimeL"/>
    </node>    
    <node pkg="ekf_loam" 
          type="IntegrationTransform"    
          name="IntegrationTransform2"    
          output="screen">
  	<remap from="/ekf_loam/integrated_to_init" to="/ekf_loam2/integrated_to_init"/>
  	<remap from="/ekf_loam/aft_mapped_to_init" to="/ekf_loam2/aft_mapped_to_init"/>
  	<remap from="/ekf_loam/filter_odom_to_initOut" to="/ekf_loam2/filter_odom_to_initOut"/>
  	<remap from="/ekf_loam/laser_odom_to_init" to="/ekf_loam2/laser_odom_to_init"/>
    </node>  
	
   <!--- Run Rviz  -->
   <node if="$(arg rviz_p)" pkg="rviz" type="rviz" name="rviz_lidar2" args="-d $(find ekf_loam)/rviz/ekf_loam_trajectory.rviz" /> 

   <!--node pkg="ekf_loam" type="data_register.py" name="data_register" args="coro" output="screen"/-->

</launch>
